impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -546,8 +543,10 @@ public boolean hasNext() {
         
         final Map.Entry<K, OnHeapValueHolder<V>> thisEntry = next;
         advance();
-    
-        setAccessTimeAndExpiry(thisEntry.getKey(), thisEntry.getValue(), timeSource.getTimeMillis());
+
+        if (setAccessTimeAndExpiryThenReturnMapping(thisEntry.getKey(), thisEntry.getValue(), timeSource.getTimeMillis(), true) == null) {
+          return null;
```

Comment:
You cannot return null, this breaks the iterator contract. We indeed clean the mapping from the store in case of failure, but we still return it.